pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+)
+
+const (
+	incrementStep29IPBlock = byte(8)
+)
+
+// IPAllocator struct maintains the current list of reserved /29 IPs pending a write operation in the cloud service. An IP Range is freed up from this list when it is available at the cloud provider end
+// As the create instance operation is in parallel, it is possible that an IP range was reserved by the service instance but is not updated in the cloud instance. The reserved list stores such IP ranges
+type IPAllocator struct {
+	reserved map[string]bool
+}
+
+// NewIPAllocator is the constructor to initialize the IPAllocator object
+// Argument reserved map[string]bool is a set of currently reserved IPs which have not been updated by the cloud provider
+func NewIPAllocator(reserved map[string]bool) *IPAllocator {
+	return &IPAllocator{
+		reserved: reserved,
+	}
+}
+
+// Reserve function adds a particular IP block in the reserve pool
+// Argument ipBlock string is an IPV4 range which needs to be reserved
+func (ipAllocator *IPAllocator) Reserve(ipV4Range string) {
+	ipAllocator.reserved[ipV4Range] = true
```

Comment:
should these be mutex?

Reply:
the  function that calls this is holding the lock